nixos/modules/services/databases/postgresql.nix

```
@@ -428,7 +454,36 @@ in
           PostgreSQL superuser account to use for various operations. Internal since changing
           this value would lead to breakage while setting up databases.
         '';
-        };
+      };
+
+      upgradeFrom = mkOption {
+        type = types.listOf (types.submodule {
+          options = {
+            package = mkOption {
+              type = types.package;
+              description = lib.mdDoc "Package to upgrade from";
+            };
+            dataDir = mkOption {
+              type = types.str;
+              example = "/var/lib/postgresql/11";
+              description = lib.mdDoc
+                "Usually in format `/var/lib/postgresql/<version>`";
+            };
+          };
+        });
+        default = [];
+        description = lib.mdDoc
+          "List of postgres versions we support upgrading from";
+      };
```

@tm-drtina Because this always upgrades to the final data directory, should this be a single package/dir pair?

I am not as familiar as you why multiple might be desired.
```suggestion
      upgradeFrom.package = mkOption {
        type = types.package;
        description = lib.mdDoc "Package to upgrade from";
      };
      upgradeFrom.dataDir = mkOption {
        type = types.str;
        example = "/var/lib/postgresql/11";
        description = lib.mdDoc
          "Usually in format `/var/lib/postgresql/<version>`";
       };
```

---

I think the idea was to support multiple machines, each possibly running a different postgres version. This way, we could upgrade all of them to the same version with a single configuration. For a single machine, it might be confusing.

To be honest, I'd also prefer a single package & directory to upgrade from. If we decide to keep the list version, I'd add more description why that's useful and how to use it when upgrading just a single machine.